File: test-fixtures/mock-mcp/spec/schema.ts
/**
 * MockMcpSpec — the single declarative behavior contract that BOTH mock impls
 * (Node `node/` and the standalone Rust crate `rust/`) deserialize, so they stay
 * behaviorally equivalent. Presets live in `spec/presets/*.json`.
 *
 * A spec is supplied to a mock via the `MOCK_MCP_SPEC` env var, which is EITHER
 * an absolute path to a preset JSON file OR an inline JSON string (impls try to
 * read it as a file first, then fall back to parsing it as a literal).
 *
 * This mirrors the host contract the desktop's Rust MCP host expects
 * (`src-tauri/src/platform/mcp_registry.rs`, `src-tauri/src/domain/mcp.rs`):
 *  - `tools/list` → `{ name, description, inputSchema }[]`
 *  - `tools/call` → a result whose flattened content drives the host's
 *    auth-signal classification (`AUTH_ERROR_RESULT_SIGNALS`) and provider
 *    extraction (`for provider '<x>'`).
 */

/** One canned outcome for a tool call. The Nth call to a tool uses
 *  `responses[min(N, responses.length - 1)]` (last response repeats). */
export interface MockToolResponse {
  /**
   * - `result`    — a normal success result (`isError=false`).
   * - `error`     — a NON-auth error body (`isError=true`); the host must NOT
   *                 misclassify it as auth (negative guard).
   * - `authError` — an auth-failure the host should classify as AuthRequired.
   */
  kind: 'result' | 'error' | 'authError';
  /** Delay before responding, ms — probes the host's per-call timeout. */
  callDelayMs?: number;
  /** Result/error body text (for `result` / `error`). */
  content?: string;
  /**
   * For `authError`: an auth signal substring to embed in the body so the host's
   * `is_auth_error_signal` matches (e.g. `"401"`, `"invalid_grant"`,
   * `"token expired"`). One of the desktop's `AUTH_ERROR_RESULT_SIGNALS`.
   */
  signal?: string;
  /**
   * For `authError`: emit `failed to fetch required token for provider '<x>': …`
   * so the host's `service_from_error_text` extracts `<x>` as the service.
   */
  providerInError?: string;
  /**
   * For `error` / `authError`: how the failure reaches the host.
   * - `isError`     — return a CallToolResult with `isError=true` (default).
   * - `raisedError` — raise a JSON-RPC/protocol error (host sees `Err(..)`).
   */
  channel?: 'isError' | 'raisedError';
}

/** A tool exposed by the mock server. */
export interface MockTool {
  name: string;
  description?: string;
  /** JSON-Schema `inputSchema` object; defaults to an empty object schema. */
  inputSchema?: Record<string, unknown>;
  /** Ordered canned outcomes (see {@link MockToolResponse}). */
  responses: MockToolResponse[];
}

/** Behavior of the `<command> auth [--provider <svc>]` re-auth subprocess. */
export interface MockAuthBehavior {
  /** Exit code when the provider check passes (0 = success). */
  exitCode?: number;
  /** When set, `auth` exits non-zero unless `--provider` equals this value. */
  requireProvider?: string;
}

/** The full mock behavior spec. */
export interface MockMcpSpec {
  /** Advertised server name (cosmetic). */
  serverName?: string;
  /** Delay the initialize handshake, ms — probes the host's startup timeout. */
  startupDelayMs?: number;
  /** Tools exposed via `tools/list` + answered by `tools/call`. */
  tools: MockTool[];
  /** `auth` subcommand behavior. */
  auth?: MockAuthBehavior;
}

const RESPONSE_KINDS = ['result', 'error', 'authError'] as const;

/**
 * Validate + normalize a parsed spec, throwing on malformed input so a bad
 * preset fails loudly rather than silently mis-behaving. Returns the spec with
 * defaults applied (empty `args`/`env`-style omissions handled by callers).
 */
export function validateSpec(raw: unknown): MockMcpSpec {
  if (!raw || typeof raw !== 'object') throw new Error('spec must be a JSON object');
  const spec = raw as Record<string, unknown>;
  if (!Array.isArray(spec.tools)) throw new Error('spec.tools must be an array');
  const tools: MockTool[] = spec.tools.map((t, i) => {
    if (!t || typeof t !== 'object') throw new Error(`spec.tools[${i}] must be an object`);
    const tool = t as Record<string, unknown>;
    if (typeof tool.name !== 'string' || !tool.name) {
      throw new Error(`spec.tools[${i}].name must be a non-empty string`);
    }
    if (!Array.isArray(tool.responses) || tool.responses.length === 0) {
      throw new Error(`spec.tools[${i}].responses must be a non-empty array`);
    }
    const responses: MockToolResponse[] = tool.responses.map((r, j) => {
      const resp = r as Record<string, unknown>;
      if (!RESPONSE_KINDS.includes(resp.kind as (typeof RESPONSE_KINDS)[number])) {
        throw new Error(`spec.tools[${i}].responses[${j}].kind invalid`);
      }
      return resp as unknown as MockToolResponse;
    });
    return {
      name: tool.name,
      description: typeof tool.description === 'string' ? tool.description : '',
      inputSchema:
        tool.inputSchema && typeof tool.inputSchema === 'object'
          ? (tool.inputSchema as Record<string, unknown>)
          : { type: 'object', properties: {} },
      responses,
    };
  });
  return {
    serverName: typeof spec.serverName === 'string' ? spec.serverName : 'mock-mcp',
    startupDelayMs: typeof spec.startupDelayMs === 'number' ? spec.startupDelayMs : 0,
    tools,
    auth: (spec.auth as MockAuthBehavior | undefined) ?? {},
  };
}

/**
 * Build the flattened error text for an `authError` response, combining the
 * provider marker (so `service_from_error_text` extracts it) and the auth signal
 * (so `is_auth_error_signal` matches). Shared by both impls' equivalents.
 */
export function authErrorText(resp: MockToolResponse): string {
  const provider = resp.providerInError
    ? `failed to fetch required token for provider '${resp.providerInError}': `
    : '';
  const signal = resp.signal ?? '401';
  const body = resp.content ?? 'authentication failed';
  return `${provider}${body} (${signal})`;
}

File: test-fixtures/mock-mcp/spec/schema.rs
//! `MockMcpSpec` — serde mirror of `schema.ts`, kept in lockstep. Both the Node
//! mock and the standalone Rust mock deserialize this same declarative contract
//! so they stay behaviorally equivalent (verified by the conformance harness).
//!
//! Included into the Rust mock crate via `#[path = "../../spec/schema.rs"] mod schema;`
//! so the schema has a single home in `spec/` shared with the TS side.

use serde::Deserialize;

/// One canned outcome for a tool call. The Nth call uses
/// `responses[min(N, len - 1)]` (the last entry repeats).
#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MockToolResponse {
    /// `"result" | "error" | "authError"`.
    pub kind: String,
    /// Delay before responding, ms — probes the host per-call timeout.
    #[serde(default)]
    pub call_delay_ms: u64,
    /// Result/error body text.
    #[serde(default)]
    pub content: Option<String>,
    /// For `authError`: an auth-signal substring to embed (e.g. `invalid_grant`).
    #[serde(default)]
    pub signal: Option<String>,
    /// For `authError`: emit `for provider '<x>'` so the host extracts the service.
    #[serde(default)]
    pub provider_in_error: Option<String>,
    /// `"isError"` (default) or `"raisedError"`.
    #[serde(default)]
    pub channel: Option<String>,
}

/// A tool exposed by the mock server.
#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MockTool {
    pub name: String,
    #[serde(default)]
    pub description: Option<String>,
    /// JSON-Schema `inputSchema` object.
    #[serde(default)]
    pub input_schema: Option<serde_json::Value>,
    pub responses: Vec<MockToolResponse>,
}

/// Behavior of the `<command> auth [--provider <svc>]` re-auth subprocess.
#[derive(Debug, Clone, Default, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MockAuthBehavior {
    /// Exit code when the provider check passes (0 = success).
    #[serde(default)]
    pub exit_code: i32,
    /// When set, `auth` exits non-zero unless `--provider` matches.
    #[serde(default)]
    pub require_provider: Option<String>,
}

/// The full mock behavior spec.
#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MockMcpSpec {
    #[serde(default)]
    pub server_name: Option<String>,
    /// Delay the initialize handshake, ms — probes the host startup timeout.
    #[serde(default)]
    pub startup_delay_ms: u64,
    pub tools: Vec<MockTool>,
    #[serde(default)]
    pub auth: MockAuthBehavior,
}

/// Build the flattened error text for an `authError` response. MUST match the
/// TypeScript `authErrorText` byte-for-byte so both mocks emit identical bodies.
pub fn auth_error_text(resp: &MockToolResponse) -> String {
    let provider = resp
        .provider_in_error
        .as_ref()
        .map(|p| format!("failed to fetch required token for provider '{p}': "))
        .unwrap_or_default();
    let signal = resp.signal.clone().unwrap_or_else(|| "401".to_string());
    let body = resp
        .content
        .clone()
        .unwrap_or_else(|| "authentication failed".to_string());
    format!("{provider}{body} ({signal})")
}
